### PR TITLE
Add Order model

### PR DESCRIPTION
**Pull request by 4Vitalii5**
User Use Cases (means these operations are allowed to users with role USER)
Use cases are the set of actions that could be performed by some actor (in this case user). Here are the list of use cases that we will cover in this part of HW:

Order Processing:

As a User, I want to place an order, so I can purchase the books in my shopping cart. I will:
Send a POST request to /api/orders to place an order.
As a User, I want to view my order history, so I can track my past purchases. I will:
Send a GET request to /api/orders to retrieve my order history.
Order Item Retrieval:

As a User, I want to view the items in my order, so I can check the details of my purchase. I will:
Send a GET request to /api/orders/{orderId}/items to retrieve all OrderItems for a specific order.
As a User, I want to view a specific item in my order, so I can check its details. I will:
Send a GET request to /api/orders/{orderId}/items/{id} to retrieve a specific OrderItem within an order.
Admin Use Cases (means these operations are allowed to users with role ADMIN)
Use cases are the set of actions that could be performed by some actor (in this case admin). Here are the list of use cases that we will cover in this part of HW:

Order Management

As an Admin, I want to update order status, so I can manage the order processing workflow. I will:
Send a PATCH request to /api/orders/{id} to update the status of an order.
Domain models (entities)
There is a list of all entities that should be present in the project:

User: Contains information about the registered user including their authentication details and personal information.
Role: Represents the role of a user in the system, for example, admin or user.
Book: Represents a book available in the store.
Category: Represents a category that a book can belong to.
ShoppingCart: Represents a user's shopping cart.
CartItem: Represents an item in a user's shopping cart.
Order: Represents an order placed by a user.
OrderItem: Represents an item in a user's order.
Order entity
Order: Represents an order placed by a user.

Add a new entity Order with the next fields:
id (Long, PK)
user (User, not null)
status (Status (enum), not null)
total (BigDecimal, not null)
orderDate (LocalDateTime, not null)
shippingAddress (String, not null)
orderItems (Set<OrderItem>)
OrderItem entity
OrderItem: Represents an item in a user's order.

Add a new entity OrderItem with the next fields:
id (Long, PK)
order (Order, not null)
book (Book, not null)
quantity (int, not null)
price (BigDecimal, not null)
General requirements
Don't forget to use Liquibase
Don't forget to implement soft delete approach
Check you are not using FetchType.EAGER in your code
Add Pagination, Sorting, and Swagger to all controllers you have
Don't forget to add validation to your input DTOs

### CODE REVIEW

File: src/main/java/mate/academy/controller/OrderController.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/orders")
+public class OrderController {
+    private final OrderService orderService;
+
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @PostMapping
+    @Operation(summary = "Place an order", description = "Make an order")
+    public OrderDto addOrder(@RequestBody @Valid CreateOrderRequestDto requestDto) {
```
Comment (JJJazl) [inline src/main/java/mate/academy/controller/OrderController.java]: Please, add the appropriate status code for POST method

File: src/main/java/mate/academy/repository/order/OrderItemRepository.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+import java.util.List;
+import java.util.Optional;
+import mate.academy.model.OrderItem;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface OrderItemRepository extends JpaRepository<OrderItem, Long> {
+    List<OrderItem> findByOrderId(Long orderId, Pageable pageable);
+
+    Optional<OrderItem> findByOrderIdAndId(Long orderId, Long itemId);
```
Comment (JJJazl) [inline src/main/java/mate/academy/repository/order/OrderItemRepository.java]:
```suggestion
    Optional<OrderItem> findByIdAndOrderId(Long itemId, Long orderId);
```

File: src/main/java/mate/academy/service/impl/OrderServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    private Order createOrderFromCart(ShoppingCart shoppingCart, CreateOrderRequestDto requestDto) {
+        BigDecimal total = calculateTotal(shoppingCart);
+        Order order = orderMapper.toOrder(shoppingCart);
+        order.setStatus(Order.Status.NEW);
+        order.setTotal(total);
+        order.setOrderDate(LocalDateTime.now());
+        order.setShippingAddress(requestDto.shippingAddress());
+        List<OrderItem> orderItems = shoppingCart.getCartItems().stream()
+                .map(cartItem -> orderItemMapper.toOrderItem(cartItem, order))
+                .toList();
```
Comment (JJJazl) [inline src/main/java/mate/academy/service/impl/OrderServiceImpl.java]:
```suggestion
        List<OrderItem> orderItems = shoppingCart.getCartItems().stream()
                .map(cartItem -> orderItemMapper.toOrderItem(cartItem))
                .map(orderItem -> orderItem.setOrder(order))
                .toList();
```

File: src/main/java/mate/academy/mapper/OrderItemMapper.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+@Mapper(config = MapperConfig.class, uses = CartItem.class)
+public interface OrderItemMapper {
+    @Mapping(source = "book.id", target = "bookId")
+    OrderItemDto toDto(OrderItem orderItem);
+
+    @Mapping(source = "book.price", target = "price")
+    @Mapping(target = "id", ignore = true)
+    @Mapping(target = "order", expression = "java(order)")
+    OrderItem toOrderItem(CartItem cartItem, @Context Order order);
```
Comment (JJJazl) [inline src/main/java/mate/academy/mapper/OrderItemMapper.java]:
```suggestion
    OrderItem toOrderItem(CartItem cartItem);
```

File: src/main/java/mate/academy/service/impl/OrderServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        orderRepository.save(order);
+        return orderMapper.toDto(order);
+    }
+
+    private Long getCurrentUserId() {
+        UserDetails principal = (UserDetails) SecurityContextHolder.getContext()
+                .getAuthentication()
+                .getPrincipal();
+        return ((User) principal).getId();
+    }
```
Comment (JJJazl) [inline src/main/java/mate/academy/service/impl/OrderServiceImpl.java]:
```suggestion
```

File: src/main/java/mate/academy/service/OrderService.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+import java.util.List;
+import mate.academy.dto.order.CreateOrderRequestDto;
+import mate.academy.dto.order.OrderDto;
+import mate.academy.dto.order.OrderItemDto;
+import mate.academy.dto.order.UpdateOrderRequestDto;
+import org.springframework.data.domain.Pageable;
+
+public interface OrderService {
+
+    OrderDto createOrder(CreateOrderRequestDto requestDto);
```
Comment (JJJazl) [inline src/main/java/mate/academy/service/OrderService.java]:
```suggestion
    OrderDto createOrder(User user, CreateOrderRequestDto requestDto);
```

File: src/main/java/mate/academy/service/OrderService.java [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+    List<OrderDto> findAllOrders(Pageable pageable);
```
Comment (JJJazl) [inline src/main/java/mate/academy/service/OrderService.java]:
```suggestion
    List<OrderDto> findAllOrders(Long userId, Pageable pageable);
```

File: src/main/java/mate/academy/controller/OrderController.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+        return orderService.createOrder(requestDto);
+    }
+
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @GetMapping
+    @Operation(summary = "Get all orders", description = "Retrieve user's order history")
+    public List<OrderDto> getOrders(Pageable pageable) {
+        return orderService.findAllOrders(pageable);
```
Comment (JJJazl) [inline src/main/java/mate/academy/controller/OrderController.java]:
```suggestion
    public List<OrderDto> getOrders(Authentication authentication, Pageable pageable) {
        User user = (User) authentication.getPrincipal();
        return orderService.findAllOrders(user.getId(), pageable);
```

File: src/main/java/mate/academy/controller/OrderController.java (hunk 1 above)
Comment (JJJazl) [inline src/main/java/mate/academy/controller/OrderController.java]:
```suggestion
    public OrderDto addOrder(Authentication authentication, @RequestBody @Valid CreateOrderRequestDto requestDto) {
```

File: src/main/java/mate/academy/controller/OrderController.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+@RestController
+@RequestMapping("/orders")
+public class OrderController {
+    private final OrderService orderService;
+
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @PostMapping
+    @Operation(summary = "Place an order", description = "Make an order")
+    public OrderDto addOrder(@RequestBody @Valid CreateOrderRequestDto requestDto) {
+        return orderService.createOrder(requestDto);
```
Comment (JJJazl) [inline src/main/java/mate/academy/controller/OrderController.java]:
```suggestion
        User user = (User) authentication.getPrincipal();
        return orderService.createOrder(user, requestDto);
```

**Review by JJJazl — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/service/impl/OrderServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+    private final OrderMapper orderMapper;
+    private final OrderItemMapper orderItemMapper;
+    private final ShoppingCartRepository shoppingCartRepository;
+
+    @Override
+    @Transactional
+    public OrderDto createOrder(User user, CreateOrderRequestDto requestDto) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(user.getId());
+        if (shoppingCart.getCartItems().isEmpty()) {
+            throw new EntityNotFoundException("Can't create order. CartItems is empty");
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/service/impl/OrderServiceImpl.java]: It's better to create 
```suggestion
            throw new OrderProcessingException("Can't create order. CartItems is empty");
```

File: src/main/java/mate/academy/service/OrderService.java [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+import java.util.List;
+import mate.academy.dto.order.CreateOrderRequestDto;
+import mate.academy.dto.order.OrderDto;
+import mate.academy.dto.order.OrderItemDto;
+import mate.academy.dto.order.UpdateOrderRequestDto;
+import mate.academy.model.User;
+import org.springframework.data.domain.Pageable;
+
+public interface OrderService {
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/service/OrderService.java]: remove empty line

File: src/main/java/mate/academy/validation/EnumValidator.java [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+        this.enumClass = annotation.enumClass();
+    }
+
+    @Override
+    public boolean isValid(String value, ConstraintValidatorContext context) {
+        if (value == null) {
+            return true;
+        }
+        return Arrays.stream(enumClass.getEnumConstants())
+                .anyMatch(e -> e.name().equals(value.toUpperCase()));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/validation/EnumValidator.java]: rename e

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
just several comments

**Review by ahoienko — APPROVED** [review]
Good job!

**Pull request merged**